Options | Connections: rewamped ui

leamas opened the pull request:
This is a re-designed UI, and also UX, for the current Options | Connections interface based on discussions in #4283

Besides the different look and feel, the major enhancement is that user gets immediate connections status feedback. The feedback is based on some new infrastructure. The driver part is implemented for serial and net-based N0183

The possibility to start the NMEA debugger from the Connections tab has been removed since there is no need. Instead the debugger can be invoked using the menu bar (Tools | NMEA Debugger), the root menu (Debug | Show NMEA Log Window) and a new 'E' hotkey.

#4300 causes several issues on current master and also on this PR.


![image](https://github.com/user-attachments/assets/6047200e-1f53-4f98-9ac5-12b4558febc4)


And: 

---

![image](https://github.com/user-attachments/assets/85712b7a-f504-4374-bb55-49a1e7bb4729)

#4300 causes several issues on current master and also on this PR

bdbcat commented:
About connections edit dialog save and persist:
Thanks to Hakan for isolating the offending commit.
Corrected by https://github.com/OpenCPN/OpenCPN/commit/dd20c0830a6d1336165d7c0992a4c9afd1abf562

leamas commented:
> About connections edit dialog save and persist: [...] Corrected by dd20c0830

...and indeed, the not-saved Enabled state in this PR is also fixed

leamas commented:
Ready to merge. Will sleep on it, though.

leamas commented:
Rebased to current master.

Please don't merge now. I have made it a habit to sleep at least a night before merging PRs. The habit has been useful...

bdbcat commented:
Oh, yes.  With something of this size, the ball is in your court.

Hakansv commented:
The connection dialog is now good and also covers all my connections. Good! More testing will come after merging.  
One minor bug in canvas menu as of the picture below.

The degraded Debug window that we've discussed earlier is back. Now much better handling on system focus. That's very good. 
But... 
This tool is primarily used when working on adding or modifying a connection. Now the Debug window cannot be started from connections when it is most needed. Nor can you change the filter or stop the flow without first exiting Options, and then back again. I still don't understand the meaning of this deterioration? Could there be some logical explanation I don't understand? 
If this is PR merged, I assume this downgrade will be there forever or is the debate still open?

![bild](https://github.com/user-attachments/assets/ffa43b85-b1cc-439a-98b8-fb6ce7d79658)

leamas commented:
In the end it is about having a consistent, well defined behaviour on all platforms vs some more clicks for windows users. It is actually no-brainer. 

The debate is always open, but you'll need someone to debate with. That person is certainly not me, I don't intend to argue about this any more

transmitterdan commented:
Is there a good argument for removing the "debug" button from the connections dialog?  Why not allow it to be started from 2 places?

transmitterdan commented:
I think there is a small bug.  If I disable a connection via left checkbox then the status does not change to "Disabled" in the Status column.  (OS-Windows 11). Even if I close the dialog and bring it back still the Status is not corrected.  If I close O and then restart then the Status finally updates.

leamas commented:
@transmitterdan 
> I think there is a small bug.

Agreed, will have a look. Might merge anyway, though.

> Is there a good argument for removing the "debug" button from the connections dialog? Why not allow it to be started from 2 places?

You are a bit late to this party, this is a very long debate. There are several reasons:
 
  - It's not logical. No user without experience would assume that the debugger is started from a connection properties page.
  - It's not consistent. All other settings on the page are subject to the Apply/Cancel/OK semantics; this setting is effective as soons as it is checked/unchecked.
  - The settings dialog is modal. According to wxwidgets specs such dialogs should block input to other windows while open. All other platforms does this, windows is the only exception. This is basically #3923. Personally, I think this is the major issue. Together with Håkan I have tried all sorts of ways to resolve this but ended up in the conclusion that there is no sane way to include this  connection properties cludge.

Now, taking a step back the core issue here is that the Settings dialog is modal. If it wasn't, it would be simple to have both the debugger and settings windows open simultaneously  and no one would ask for a fishy _please let me start the debugger from the connections properties page_ We are now heading in this direction: https://github.com/OpenCPN/OpenCPN/discussions/4308#discussioncomment-11826723

leamas commented:
>  It's the first thing they would do to see that the messages show up whilst fiddling with, for example, the serial port settings or perhaps the IP port #. It's probably the 98% use case.

It's just that they don't need to, the new status field reflects this

transmitterdan commented:
I have been around users and software developers a long time. A user wants to see the messages if they can. It's much more satisfying than a status bit. If the button goes away, users will complain. If it stays, they will use it.

The serial port is a classic example of why. If you get the port # correct but the baud rate not, there will be characters flowing but text will be gibberish. Once you have that then adjusting baud until the characters make sense is easy. The status bit won't help in this case I think.

I understand about the modal issue. I'm just seeing it from the user point of view who has no interest in the vagaries of mode/modeless debates.

The new UI is nice. It is more intuitive and cleaner. Users, however, are messy and I am continually amazed at how seemingly unintuitive they can be. But they are "right" by definition.

leamas commented:
>  understand about the modal issue. I'm just seeing it from the user point of view who has no interest in the vagaries of mode/modeless debates.

Well, you see it from a Windows user perspective, I'd say.  The result on other platforms is a more or less unusable mess. 

The way to handle this is to merge the PR as it is, it represents a sane, new  starting point working on all platforms. I am open for some kind of cludge for Windows users only on top of that, but  I'd expect you and/or Håkan to create it.  Starting with a clean sheet this would probably not be a checkbox, YMMV.

And,when @bdbcat completes the work to make the settings modeless such a cludge would be even more of a, well, cludge.

transmitterdan commented:
My comment was not meant to suggest a course of action different than what you recommend. I was just commenting on an assumption about users. Non-Windows users are atypical compared the general case.

leamas commented:
> Non-Windows users are atypical compared the general case.

We all live in different bubbles.  Defining windows as the "general case" is one. :)

leamas commented:
> The serial port is a classic example of why. If you get the port # correct but the baud rate not, there will be characters flowing but text will be gibberish. Once you have that then adjusting baud until the characters make sense is easy. The status bit won't help in this case I think

It actually does. It reflects decoded messages, not just some low-level driver status

Hakansv commented:
Not surprisingly, I agree with TDan.
"Bubbles"?
On my courses with about 15 participants in recent years, there has never been more than one Linux/Mac user, usually none. The rest are Win often combined with an Android.
Some statistics from o-charts user system.

![bild](https://github.com/user-attachments/assets/680c7c1c-7653-43d6-863f-990a871206ba)

transmitterdan commented:
> > The serial port is a classic example of why. If you get the port # correct but the baud rate not, there will be characters flowing but text will be gibberish. Once you have that then adjusting baud until the characters make sense is easy. The status bit won't help in this case I think
> 
> It actually does. It reflects decoded messages, not just some low-level driver status

If it would be the case that the status bit had more than 2 states (good/bad) then I might agree.  For serial ports there are 2 or 3 conditions one can see in the debug window.  First is to get the port right.  If that is not right, there will be no data whatsoever. The right port but wrong baud will produce gibberish text. Next is to get the baud rate right.  For that, properly formed text that includes identifiable numbers and other typical strings is present when the correct rate is selected.  Finally, the protocol must be right so the strings can be properly interpreted.  I think the current status bit only reflects whether everything is right or not. Whereas the information about each step above is readily apparent in the debug window, at least for us Neanderthal Windows users.

I am not trying to pick a fight. I'm just relaying my experience, and I think the experience of a majority of users. There are readily available debug tools such as WireShark for IP stuff. But not so much for serial ports. OpenCPN has long been a pretty good tool for debugging serial port (such as USB port) connections.  Windows has near zero help but OpenCPN helps a lot.

leamas commented:
>  First is to get the port right. If that is not right, there will be no data whatsoever.  

There is a specific status X for "wrong port" or "cannot open xxx"

>The right port but wrong baud will produce gibberish text.

This is the ! status "Data is not flowing"

> For that, properly formed text that includes identifiable numbers and other typical strings is present when the correct rate is selected. Finally, the protocol must be right so the strings can be properly interpreted.

And this is the final "Data is flowing" status


 > I think the current status bit only reflects whether everything is right or not

Obviously, not

rgleason commented:
Now that windows has been put back in the return box. We can anticipate the initial howls, but I think you're all going to find a way to make work better eventually.

TwoCanPlugIn commented:
Sorry about being late to the party.

The new look makes sense, however I disagree with the new UI metaphor.

Users will not be familiar with clicking on a an icon in a grid. Whilst my preference would be to ditch the gear  & rubbish bin icon, use the remaining space for the user comment field and have a right click context menu for delete, edit and view statistics actions, (assuming O keeps things like packets per second, error counts), I think that retaining the "Remove" and "Edit" buttons would maintain familiarity and reduce any re-learning requirements for users. I don't know of any other OS UI that uses clickable options in a grid to alter system settings.

The discussion about serial port baud rates, should be a non issue. I'm relatively confident, that on Windows at least, the baud rate for USB adapters such as the Yacht Devices YDNU-2 has no practical effect. Although unrelated to this, my own testing of USB CAN adapters on Windows such as the Canable, Peak, it doesn't matter what baud rate is set, it just works.  It is important however for traditional RS-232 devices. 

For USB adapters, users should not need to know the port number, as it can be retrieved from the registry on Windows, IOService on OSX and  either udev rules or traversing /sys/bus/usb-serial/devices on Linux to find the matching port given a PID/VID. If something can be retrieved from the system, it should be preferable to the user entering a setting. 

I'm not convinced that the upload format option belongs here. For that matter, doesn't Route/Points deserve their own Settings, rather than just being a tab on the "Ships" option ? To me, anything to do with uploading/downloading/import/export  of routes, tracks, marks, which includes the Garmin or Furuno format belongs here. And perhaps I'm being pedantic, but a "mark" is something I sail around during a yacht race, a waypoint is a destination or an intermediate part of a route.

rgleason commented:
Steve, I agree with much of what you suggest, and even your statement about marks, but we have had discussions about marks and waypoints in OpenCPN, and for me, that is a settle matter. (Just a suggestion: If you were to tone down just a little, so it could be a discussion, the message might be heard better.)

leamas commented:
Cleaned up history, fixed the issue about disabling a connection not being reflected in the state column.

@TwoCanPlugIn 

> Sorry about being late to the party.

Indeed, you are to late. Going to merge this. If you have ideas for  a better implementation please file  separate issues.

OTOH a new party is available at #4308, you are welcome there!

leamas commented:
Here we go...
